File: tools/te2rom.py
# -*- coding: utf-8 -*-
"""Telugu script -> ISO-ish romanization, matching the convention already used across
this project (ā ī ū ē ō · ṭ ḍ ṇ ḷ ṣ ś · ch for చ · ṁ for anusvara).

This direction is near-deterministic: the script encodes vowel length, retroflexion and
gemination explicitly, which is exactly what romanization needs. The reverse direction is
not, which is why it is not attempted anywhere in this project.
"""
import re, unicodedata

VIRAMA = '్'
ANUSVARA = 'ం'
VISARGA = 'ః'
CANDRABINDU = 'ఁ'

CONS = {
 'క':'k','ఖ':'kh','గ':'g','ఘ':'gh','ఙ':'ṅ',
 'చ':'ch','ఛ':'chh','జ':'j','ఝ':'jh','ఞ':'ñ',
 'ట':'ṭ','ఠ':'ṭh','డ':'ḍ','ఢ':'ḍh','ణ':'ṇ',
 'త':'t','థ':'th','ద':'d','ధ':'dh','న':'n',
 'ప':'p','ఫ':'ph','బ':'b','భ':'bh','మ':'m',
 'య':'y','ర':'r','ఱ':'ṟ','ల':'l','ళ':'ḷ','ఴ':'ḻ',
 'వ':'v','శ':'ś','ష':'ṣ','స':'s','హ':'h','క్ష':'kṣ',
}
IND_VOW = {
 'అ':'a','ఆ':'ā','ఇ':'i','ఈ':'ī','ఉ':'u','ఊ':'ū',
 'ఋ':'ṛ','ౠ':'ṝ','ఌ':'ḷ̥','ౡ':'ḹ',
 'ఎ':'e','ఏ':'ē','ఐ':'ai','ఒ':'o','ఓ':'ō','ఔ':'au',
}
SIGN_VOW = {
 'ా':'ā','ి':'i','ీ':'ī','ు':'u','ూ':'ū',
 'ృ':'ṛ','ౄ':'ṝ','ె':'e','ే':'ē','ై':'ai',
 'ొ':'o','ో':'ō','ౌ':'au','ౢ':'l̥','ౣ':'l̥̄',
}
DIGITS = {chr(0x0c66+i): str(i) for i in range(10)}


ZERO_WIDTH = '\u200c\u200d\ufeff'   # ZWNJ/ZWJ appear inside loanwords: డాక్టర్‌ని

def romanize_word(w):
    w = ''.join(c for c in w if c not in ZERO_WIDTH)
    out = []
    i, n = 0, len(w)
    while i < n:
        c = w[i]
        if c in IND_VOW:
            out.append(IND_VOW[c]); i += 1; continue
        if c in DIGITS:
            out.append(DIGITS[c]); i += 1; continue
        if c in CONS:
            out.append(CONS[c]); i += 1
            # a consonant carries an inherent 'a' unless a vowel sign or virama follows
            if i < n and w[i] == VIRAMA:
                i += 1                      # dead consonant, no vowel
            elif i < n and w[i] in SIGN_VOW:
                out.append(SIGN_VOW[w[i]]); i += 1
            else:
                out.append('a')
            continue
        if c == ANUSVARA:
            out.append('ṁ'); i += 1; continue
        if c == VISARGA:
            out.append('ḥ'); i += 1; continue
        if c == CANDRABINDU:
            out.append('m̐'); i += 1; continue
        if c in SIGN_VOW:                    # stray sign, shouldn't happen
            out.append(SIGN_VOW[c]); i += 1; continue
        out.append(c); i += 1                # punctuation, latin, spaces
    return ''.join(out)


def _anusvara_place(s):
    """Anusvara assimilates to the following stop. The project writes the result rather
    than a bare ṁ: తింటాను -> tinṭānu, ఇంగ్లీష్ -> ingliṣ."""
    out = []
    for i, ch in enumerate(s):
        if ch != 'ṁ':
            out.append(ch); continue
        nxt = s[i+1] if i+1 < len(s) else ''
        # NB: '' is a substring of every string, so the empty case must be tested first
        if not nxt or nxt.isspace():  out.append('ṁ')   # word-final stays ṁ
        # ṅ and ñ carry no information an English reader lacks — "ng", "nch" already read
        # correctly — so both are written n, which is also what the course pages mostly do.
        elif nxt in 'kg':             out.append('n')   # rangu, inkā
        elif nxt in 'cj':             out.append('n')   # manchi, sanjay
        elif nxt in 'ṭḍ':             out.append('ṇ')   # reṇḍu, iṇṭlō — retroflexion does matter
        elif nxt in 'tdn':            out.append('n')
        elif nxt in 'pbm':            out.append('m')
        else:                         out.append('ṁ')   # before a fricative, y, l, v…
    return ''.join(out)


def romanize(text, assimilate=True):
    parts = re.split(r'(\s+)', text.strip())
    out = ''.join(p if p.isspace() else romanize_word(p) for p in parts)
    return _anusvara_place(out) if assimilate else out


if __name__ == '__main__':
    import json, sys, os
    D = os.path.dirname(os.path.abspath(__file__))
    pairs = []
    for fn in ('anki_rows.json', 'sent_rows.json'):
        p = os.path.join(D, fn)
        if not os.path.exists(p): continue
        for r in json.load(open(p)):
            rom, tel = r.get('rom', ''), r.get('tel', '')
            if rom and tel and not any(c in rom for c in '→·/_()[]'):
                pairs.append((tel, rom))
    seen, uniq = set(), []
    for t, r in pairs:
        if (t, r) not in seen: seen.add((t, r)); uniq.append((t, r))
    ok, bad = 0, []
    for tel, gold in uniq:
        got = romanize(tel)
        if got.lower() == gold.lower(): ok += 1
        else: bad.append((tel, gold, got))
    print(f'attested pairs: {len(uniq)}   exact: {ok}  ({100*ok/len(uniq):.1f}%)')
    for b in bad[:40]:
        print(f'   {b[0]:<22} gold={b[1]:<24} got={b[2]}')

File: STUDY/assets/te2rom.js
/* Telugu script -> the project's romanization. A direct port of tools/te2rom.py.
 *
 * WHY A SECOND IMPLEMENTATION EXISTS, AND HOW IT IS KEPT HONEST
 * The Python version runs in the offline pipeline. This one runs in the browser, for
 * transcripts that cannot be committed and therefore cannot be baked — without it those texts
 * render as script the learner cannot read, which defeats the point of loading them.
 *
 * Two implementations of the same rule is exactly the drift this project has been careful to
 * avoid, so the port is deliberately mechanical — same tables, same order of tests, same
 * assimilation rules — and `tools/check_te2rom.py` diffs the two over every distinct Telugu
 * token available. Change one, run that, or they will part company silently.
 *
 * The direction is near-deterministic: the script marks vowel length, retroflexion and
 * gemination explicitly, which is what the romanization needs. The reverse is not, and is not
 * attempted anywhere in this project.
 */
const Te2Rom = (() => {
  const VIRAMA = '్', ANUSVARA = 'ం', VISARGA = 'ః', CANDRABINDU = 'ఁ';

  const CONS = {
    'క': 'k', 'ఖ': 'kh', 'గ': 'g', 'ఘ': 'gh', 'ఙ': 'ṅ',
    'చ': 'ch', 'ఛ': 'chh', 'జ': 'j', 'ఝ': 'jh', 'ఞ': 'ñ',
    'ట': 'ṭ', 'ఠ': 'ṭh', 'డ': 'ḍ', 'ఢ': 'ḍh', 'ణ': 'ṇ',
    'త': 't', 'థ': 'th', 'ద': 'd', 'ధ': 'dh', 'న': 'n',
    'ప': 'p', 'ఫ': 'ph', 'బ': 'b', 'భ': 'bh', 'మ': 'm',
    'య': 'y', 'ర': 'r', 'ఱ': 'ṟ', 'ల': 'l', 'ళ': 'ḷ', 'ఴ': 'ḻ',
    'వ': 'v', 'శ': 'ś', 'ష': 'ṣ', 'స': 's', 'హ': 'h',
  };
  const IND_VOW = {
    'అ': 'a', 'ఆ': 'ā', 'ఇ': 'i', 'ఈ': 'ī', 'ఉ': 'u', 'ఊ': 'ū',
    'ఋ': 'ṛ', 'ౠ': 'ṝ', 'ఌ': 'ḷ̥', 'ౡ': 'ḹ',
    'ఎ': 'e', 'ఏ': 'ē', 'ఐ': 'ai', 'ఒ': 'o', 'ఓ': 'ō', 'ఔ': 'au',
  };
  const SIGN_VOW = {
    'ా': 'ā', 'ి': 'i', 'ీ': 'ī', 'ు': 'u', 'ూ': 'ū',
    'ృ': 'ṛ', 'ౄ': 'ṝ', 'ె': 'e', 'ే': 'ē', 'ై': 'ai',
    'ొ': 'o', 'ో': 'ō', 'ౌ': 'au', 'ౢ': 'l̥', 'ౣ': 'l̥̄',
  };
  const DIGITS = {};
  for (let i = 0; i < 10; i++) DIGITS[String.fromCharCode(0x0c66 + i)] = String(i);

  const ZERO_WIDTH = '‌‍﻿';   // ZWNJ/ZWJ turn up inside loanwords

  function romanizeWord(w) {
    w = [...w].filter(c => !ZERO_WIDTH.includes(c)).join('');
    const out = [];
    let i = 0;
    const n = w.length;
    while (i < n) {
      const c = w[i];
      if (IND_VOW[c] !== undefined) { out.push(IND_VOW[c]); i++; continue; }
      if (DIGITS[c] !== undefined) { out.push(DIGITS[c]); i++; continue; }
      if (CONS[c] !== undefined) {
        out.push(CONS[c]); i++;
        // a consonant carries an inherent 'a' unless a vowel sign or virama follows
        if (i < n && w[i] === VIRAMA) i++;
        else if (i < n && SIGN_VOW[w[i]] !== undefined) { out.push(SIGN_VOW[w[i]]); i++; }
        else out.push('a');
        continue;
      }
      if (c === ANUSVARA) { out.push('ṁ'); i++; continue; }
      if (c === VISARGA) { out.push('ḥ'); i++; continue; }
      if (c === CANDRABINDU) { out.push('m̐'); i++; continue; }
      if (SIGN_VOW[c] !== undefined) { out.push(SIGN_VOW[c]); i++; continue; }
      out.push(c); i++;                         // punctuation, latin, spaces
    }
    return out.join('');
  }

  /* Anusvara assimilates to the following stop, and the project writes the result rather than
     a bare ṁ: తింటాను -> tiṇṭānu. ṅ and ñ carry nothing an English reader lacks — "ng" and
     "nch" already read correctly — so both are written n. Retroflexion does matter, so ṇ
     survives before ṭ and ḍ. */
  function anusvaraPlace(s) {
    const chars = [...s];
    return chars.map((ch, i) => {
      if (ch !== 'ṁ') return ch;
      const nxt = chars[i + 1] || '';
      if (!nxt || /\s/.test(nxt)) return 'ṁ';   // word-final stays ṁ
      if ('kg'.includes(nxt)) return 'n';
      if ('cj'.includes(nxt)) return 'n';
      if ('ṭḍ'.includes(nxt)) return 'ṇ';
      if ('tdn'.includes(nxt)) return 'n';
      if ('pbm'.includes(nxt)) return 'm';
      return 'ṁ';                                // before a fricative, y, l, v…
    }).join('');
  }

  function romanize(text, assimilate = true) {
    const parts = String(text || '').trim().split(/(\s+)/);
    const out = parts.map(p => /^\s+$/.test(p) ? p : romanizeWord(p)).join('');
    return assimilate ? anusvaraPlace(out) : out;
  }

  return { romanize, romanizeWord };
})();
